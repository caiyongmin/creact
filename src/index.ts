import React, {
  useState,
  useEffect,
  useReducer,
  useCallback,
  useMemo,
  useRef,
  useContext,
} from './react/index';
import ReactDOM from './react-dom/index';

export {
  React,
  ReactDOM,
  useState,
  useEffect,
  useReducer,
  useCallback,
  useMemo,
  useRef,
  useContext,
}
