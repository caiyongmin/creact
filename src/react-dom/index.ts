import { render } from './render';
import { diff } from './diff';

export default {
  render,
  diff
};
