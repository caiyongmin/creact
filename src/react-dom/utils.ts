import { VNode } from "../../typings/index";

export function isStrOrNum(val: any): val is string | number {
  return typeof val === 'string' || typeof val === 'number';
}

export function isUnRenderVDom(val: any): val is boolean | undefined | null {
  return typeof val === 'boolean' || val === undefined || val === null;
}

export function isObject(val: any): boolean {
  return typeof val === 'object' && val !== null;
}

export function isFunction(val: any): val is Function {
  return typeof val === 'function';
}

export function getFunctionBody (fn: Function): string {
  const toString = Function.prototype.toString;
  return toString.call(fn).replace(/^[^{]*{\s*/, '').replace(/\s*}[^}]*$/, '')
}

export function isClass(val: any): boolean {
  if (typeof val !== 'function') {
    return false;
  }

  if (/^class[\s{]/.test(val.toString())) {
    return true;
  }

  const fnBody = getFunctionBody(val);
  const result = (/classCallCheck/.test(fnBody)
    || /TypeError\("Cannot call a class as a function"\)/.test(fnBody));

  return result;
}

export function isString(val: any): val is string {
  return typeof val === 'string';
}

export function isSameNodeType(dom: any, vdom: VNode): boolean {
  // 文本节点
  if (typeof vdom === 'string' || typeof vdom === 'number') {
    return dom.nodeType === 3;
  }

  // DOM 节点
  if (typeof vdom === 'object' && vdom !== null && typeof vdom.type === 'string') {
    return dom.nodeName.toLowerCase() === vdom.type.toLowerCase();
  }

  // 组件
  if (dom && dom.__componentInstance && typeof vdom === 'object' && vdom !== null && vdom.type) {
    return dom.__componentInstance.renderVDOM === vdom.type;
  }

  return false;
}

export function shallowCompareObject(
  obj: { [ key: string ]: any },
  compareObj: { [ key: string ]: any },
) {
  if (!isObject(obj) || !isObject(compareObj)) {
    return false;
  }

  const objKeys = Object.keys(obj);
  const compareObjKeys = Object.keys(compareObj);

  if (objKeys.length !== compareObjKeys.length) {
    return false;
  }

  let index = objKeys.length;
  while (index--) {
    const key = objKeys[index];

    if (
      !Object.hasOwnProperty.call(compareObj, key)
      || compareObj[key] !== obj[key]
    ) {
      return false;
    }
  }

  return true;
}
