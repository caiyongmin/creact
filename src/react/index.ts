import Component from './component';
import createElement from './createElement';

export { setDispatcher, useState, useEffect } from './hooks';

export default {
  Component,
  createElement
}
