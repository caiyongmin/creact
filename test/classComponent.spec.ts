import { createRoot, deleteRoot, sleep } from './utils';
import { ReactDOM, React } from './../src';
import { DemoComponentVNode } from './fixtures/classComponent';

let lifeCycleCall = '';
const TriggerComponent = class Trigger extends React.Component<{ count: number }, { num: number }> {
  constructor(props: { count: number }) {
    super(props);

    this.state = {
      num: 1,
    };
  }

  componentWillMount() {
    lifeCycleCall += 'componentWillMount;';
  }

  componentDidMount() {
    lifeCycleCall += 'componentDidMount;';
  }

  componentWillReceiveProps() {
    lifeCycleCall += 'componentWillReceiveProps;';
  }

  shouldComponentUpdate() {
    lifeCycleCall += 'shouldComponentUpdate;';
    return true;  // always return true, to going through an update process
  }

  componentWillUpdate() {
    lifeCycleCall += 'componentWillUpdate;';
  }

  componentDidUpdate() {
    lifeCycleCall += 'componentDidUpdate;';
  }

  componentWillUnmount() {
    lifeCycleCall += 'componentWillUnmount;';
  }

  handleAddNum = () => {
    this.setState({
      num: this.state.num + 1,
    });
  }

  render() {
    lifeCycleCall += 'render;';
    return React.createElement(
      "div",
      {},
      React.createElement("button", { id: 'num', onClick: this.handleAddNum }, "addNum"),
      React.createElement("button", { id: 'force-update', onClick: () => this.forceUpdate() }, "forceUpdate"),
      React.createElement("span", { id: 'date' }, +new Date()),
      React.createElement("span", {}, this.props.count),
    )
  }
};

// 父组件，内含类型为 classComponent 的子组件
// 父组件主要用来为子组件提供需要的 props 参数
const VNode = React.createElement(
  class Demo extends React.Component<{}, { count: number, toggle: boolean }> {
    constructor(props: any) {
      super(props);

      this.state = {
        count: 0,
        toggle: false,
      };
    }

    handleAdd = () => {
      this.setState({
        count: this.state.count + 1,
      });
    }

    handleToggle = () => {
      this.setState({
        toggle: !this.state.toggle,
      });
    }

    render() {
      return React.createElement(
        'div',
        {},
        React.createElement("button", { id: 'add', onClick: this.handleAdd }, "addCount"),
        React.createElement("button", { id: 'toggle', onClick: this.handleToggle }, "toggle"),
        this.state.toggle
        ? React.createElement('span', {}, String(this.state.toggle))
        : React.createElement(
          // 类型为 classComponent 的子组件
          TriggerComponent,
          { id: 'result', count: this.state.count, }
        )
      )
    }
  }
);

describe('class component', () => {
  let root: HTMLDivElement;

  beforeEach(() => {
    root = createRoot();
  });

  afterEach(() => {
    deleteRoot(root);
  });

  it('should render correct DOM', () => {
    ReactDOM.render(DemoComponentVNode, root);
    expect(root).toMatchSnapshot();
  });

  it('should render correct DOM after virtual dom diff', async () => {
    ReactDOM.render(DemoComponentVNode, root);
    const addButton = root.querySelector('#add') as HTMLButtonElement;
    const result = root.querySelector('#result') as HTMLSpanElement;
    expect(result.innerHTML).toBe('0');
    addButton.click();
    await sleep(50);
    expect(result.innerHTML).toBe("1");
  });

  it('should setState is defer', async () => {
    ReactDOM.render(DemoComponentVNode, root);
    const addButton = root.querySelector('#add') as HTMLButtonElement;
    const result = root.querySelector('#result') as HTMLSpanElement;
    expect(result.innerHTML).toBe('0');
    addButton.click();  // want add 1
    addButton.click();  // want add 1
    addButton.click();  // want add 1
    expect(result.innerHTML).toBe('0');  // but result value still 0
    await sleep(50);
    expect(result.innerHTML).toBe('1');  // state value is 1
  });

  it('should work about setState with stateUpdater', async () => {
    ReactDOM.render(DemoComponentVNode, root);
    const stateUpdaterButton = root.querySelector('#state-updater') as HTMLButtonElement;
    const result = root.querySelector('#result') as HTMLSpanElement;
    expect(result.innerHTML).toBe('0');
    stateUpdaterButton.click();  // want add 1
    stateUpdaterButton.click();  // want add 1
    stateUpdaterButton.click();  // want add 1
    expect(result.innerHTML).toBe('0');
    await sleep(50);
    expect(result.innerHTML).toBe('3');
    expect(0).toBe(0);
  });

  it('should call the correct life cycle function', async () => {
    ReactDOM.render(VNode, root);
    const addButton = root.querySelector('#add') as HTMLButtonElement;
    const forceUpdateButton = root.querySelector('#force-update') as HTMLButtonElement;
    const toggleButton = root.querySelector('#toggle') as HTMLButtonElement;
    const numButton = root.querySelector('#num') as HTMLSpanElement;

    // mounting
    expect(lifeCycleCall).toBe(''
      + 'componentWillMount;'
      + 'render;'
      + 'componentDidMount;'
    );

    // updating from new props
    lifeCycleCall = '';
    addButton.click();
    await sleep(50);
    expect(lifeCycleCall).toBe(''
      + 'componentWillReceiveProps;'
      + 'shouldComponentUpdate;'
      + 'componentWillUpdate;'
      + 'render;'
      + 'componentDidUpdate;'
    );

    // updating from new state
    lifeCycleCall = '';
    numButton.click();
    await sleep(50);
    expect(lifeCycleCall).toBe(''
      + 'shouldComponentUpdate;'
      + 'componentWillUpdate;'
      + 'render;'
      + 'componentDidUpdate;'
    );

    // updating from forceUpdate
    lifeCycleCall = '';
    forceUpdateButton.click();
    await sleep(50);
    expect(lifeCycleCall).toBe(''
      + 'componentWillUpdate;'
      + 'render;'
      + 'componentDidUpdate;'
    );

    // unmounting
    lifeCycleCall = '';
    toggleButton.click();
    await sleep(50);
    expect(lifeCycleCall).toBe(''
      + 'componentWillUnmount;'
    );
  });
});
