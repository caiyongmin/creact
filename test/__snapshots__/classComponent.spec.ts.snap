// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class component should render correct DOM 1`] = `
<div
  id="root"
>
  <div>
    <button
      id="add"
    >
      addCount
    </button>
    <button
      id="state-updater"
    >
      stateUpdater
    </button>
    <span
      id="result"
    >
      0
    </span>
  </div>
</div>
`;
